src/main/java/net/neoforged/neoforge/resource/ContextAwareReloadListener.java

```diff
@@ -5,15 +5,67 @@
 
 package net.neoforged.neoforge.resource;
 
+import com.google.gson.JsonElement;
+import com.mojang.serialization.JsonOps;
+import java.lang.ref.WeakReference;
 import net.minecraft.core.RegistryAccess;
+import net.minecraft.resources.RegistryOps;
+import net.minecraft.server.packs.resources.PreparableReloadListener;
+import net.minecraft.server.packs.resources.SimplePreparableReloadListener;
+import net.neoforged.neoforge.common.conditions.ConditionalOps;
 import net.neoforged.neoforge.common.conditions.ICondition;
+import net.neoforged.neoforge.common.conditions.ICondition.IContext;
+import org.jetbrains.annotations.ApiStatus;
 
-public abstract class ContextAwareReloadListener {
+/**
+ * Reload listeners that descend from this class will have the reload context automatically populated when it is available.
+ * <p>
+ * The context is guaranteed to be available for the duration of {@link PreparableReloadListener#reload}.
+ * <p>
+ * For children of {@link SimplePreparableReloadListeners}, it will be available during both {@link SimplePreparableReloadListener#prepare prepare()} and {@link SimplePreparableReloadListener#apply apply()}.
+ */
+public abstract class ContextAwareReloadListener implements PreparableReloadListener {
+    /**
+     * @deprecated Use {@link #getContext()}
+     */
+    @Deprecated(forRemoval = true)
     protected ICondition.IContext conditionContext = ICondition.IContext.EMPTY;
+
+    /**
+     * @deprecated Use {@link #getRegistryAccess()}
+     */
+    @Deprecated(forRemoval = true)
     protected RegistryAccess registryAccess = RegistryAccess.EMPTY;
 
-    public void injectContext(ICondition.IContext conditionContext, RegistryAccess registryAccess) {
-        this.conditionContext = conditionContext;
-        this.registryAccess = registryAccess;
+    private WeakReference<ICondition.IContext> context = new WeakReference<>(null);
```

Comment:
Why the weak references? Vanilla stores the registry access in the server advancement manger directly in recent snapshots, and everything is already retained via ServerReloadableResources.

Additionally, weak references will just cause the context to get GC'ed too early if everything is moved to them.